### PR TITLE
SipHash: add separate setter for the hash size

### LINKED ISSUE #7143: Crash when try to set siphash's hash size to 8 through the EVP interface
in 1.1.1-pre9

I call following code:
```
EVP_PKEY * pkey = EVP_PKEY_new_raw_private_key (EVP_PKEY_SIPHASH, NULL, key, 16);
EVP_MD_CTX * mdctx = EVP_MD_CTX_create ();	
EVP_PKEY_CTX * ctx = NULL;
EVP_DigestSignInit (mdctx, &ctx, NULL, NULL, pkey);
EVP_PKEY_CTX_ctrl (ctx, -1, EVP_PKEY_OP_SIGNCTX, EVP_PKEY_CTRL_SET_DIGEST_SIZE, 8, NULL));
```	

It passes null pointer here
https://github.com/openssl/openssl/blob/master/crypto/siphash/siphash_pmeth.c#L139
and crashes here
https://github.com/openssl/openssl/blob/master/crypto/siphash/siphash.c#L86
because of the  null pointer for k.

### PR DESCRIPTION
**Pull request by levitte**
This was originally part of SipHash_Init.  However, there are cases
where there isn't any key material to initialize from when setting the
hash size, and we do allow doing so with a EVP_PKEY control.  The
solution is to provide a separate hash_size setter and to use it in
the corresponding EVP_PKEY_METHOD.

Fixes #7143

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by mattcaswell** [comment]
CI failures look relevant.

**Comment by tmshort** [comment]
Looks reasonable; but what are the Travis errors?

**Comment by levitte** [comment]
I found a fault (when the hash size wasn't set, it remained zero).  Fixed.
Pushed that before I saw the CI, so I'll wait for the re-run to complete.

**Comment by levitte** [comment]
Travis happier now...

**Comment by t-j-h** [comment]
You aren't testing usage without calling set_hash_size in the test case ...

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Comment by mattcaswell** [comment]
Can we have a test somewhere that sets the hash size to 8 via EVP as per the scenario that triggered this?

File: crypto/siphash/siphash_pmeth.c [hunk 1]
```diff
@@ -135,8 +131,7 @@ static int pkey_siphash_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
             p1 != SIPHASH_MAX_DIGEST_SIZE) {
             return 0;
         }
```
[the remark this answers is not in the document]
Comment (levitte) [reply crypto/siphash/siphash_pmeth.c]: Ah, good point...  I think.  Will check

**Comment by levitte** [comment]
> Can we have a test somewhere that sets the hash size to 8 via EVP as per the scenario that triggered this?

Sure, that should be possible (actually, exactly what's needed...   such a test would have shown this bug long ago)

File: crypto/siphash/siphash.c [hunk 2]
```diff
@@ -80,12 +80,18 @@ size_t SipHash_hash_size(SIPHASH *ctx)
     return ctx->hash_size;
 }
 
-int SipHash_set_hash_size(SIPHASH *ctx, size_t hash_size)
+static size_t siphash_adjust_hash_size(size_t hash_size)
 {
     if (hash_size == 0)
         hash_size = SIPHASH_MAX_DIGEST_SIZE;
-    else if (hash_size != SIPHASH_MIN_DIGEST_SIZE &&
-             hash_size != SIPHASH_MAX_DIGEST_SIZE)
```
Comment (orignal) [inline crypto/siphash/siphash.c]: hash size must be either 8 or 16. It can't be, for example, 10. But this changes makes it allowed

Comment (tmshort) [reply crypto/siphash/siphash.c]: There is a unit-test for size 4 to error out, as far as I know that still passes.

**Review by t-j-h — APPROVED** [review]

File: crypto/siphash/siphash.c (hunk 2 above)
Comment (levitte) [reply crypto/siphash/siphash.c]: `SipHash_set_hash_size` does the check and returns 0 when `hash_size` isn't one of those values, so I'm not sure what code path you're talking about.

Comment (orignal) [reply crypto/siphash/siphash.c]: never mind. It's checked in another place now

**Review by t-j-h — APPROVED** [review]

**Comment by levitte** [comment]
Merged.

d74f23d2dbf2de0f374bff004c135242cfb65174 SipHash: add separate setter for the hash size
2725232132207ee30fd16e596885e3b0bd810fc4 TESTS: add SipHash tests with digestsize controls
d0123525cd82be320f6b8b91ca287a2e572341cc SipHash: make it possible to control the hash size through string controls
7e6a3025493bae3018e8a9c71b93350859b1c484 test/evp_test.c: make it possible to use controls with MAC tests

**Pull request closed without merging**